pkg/windows/calico.go

```diff
@@ -52,147 +53,226 @@ const (
+	logrus.Info("Generating HNS networks, please wait")
+	if err := c.generateCalicoNetworks(); err != nil {
+		return err
 	}
-	calicoKubeConfig.Token = token.Status.Token
 
-	cfg := &CNIConfig{NodeConfig: nodeConfig, NetworkName: "Calico", BindAddress: nodeConfig.AgentConfig.NodeIP}
+	return nil
+}
 
-	hc, err := helm.NewFactoryFromConfig(restConfig)
+func (c *Calico) GetConfig() *CNIConfig {
+	return c.CalicoCNICfg
+}
+
+
+// initializeConfig sets the default configuration in CNIConfig
+func (c *Calico) initializeConfig(ctx context.Context, nodeConfig *daemonconfig.Node, restConfig *rest.Config) error {
+	c.CalicoCNICfg = &CNIConfig{NetworkName: "Calico", BindAddress: nodeConfig.AgentConfig.NodeIP}
+
+	platformType, err := getPlatformType()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if err := getDefaultConfig(cfg, dataDir, nodeConfig); err != nil {
-		return nil, err
+	c.CalicoCNICfg.CalicoConfig = &CalicoConfig{
+		Name:                  "Calico",
+		Hostname:              nodeConfig.AgentConfig.NodeName,
+		NodeNameFile:          filepath.Join("c:\\", c.DataDir, "agent", CalicoNodeNameFileName),
+		KubeNetwork:           "Calico.*",
+		Mode:                  "vxlan",
+		ServiceCIDR:           nodeConfig.AgentConfig.ServiceCIDR.String(),
+		DNSServers:            nodeConfig.AgentConfig.ClusterDNS.String(),
+		DNSSearch:             "svc." + nodeConfig.AgentConfig.ClusterDomain,
+		DatastoreType:         "kubernetes",
+		NetworkingBackend:     "vxlan",
+		Platform:              platformType,
+		StartUpValidIPTimeout: 90,
+		LogDir:                "",
+		IP:                    "autodetect",
+		IPAutoDetectionMethod: "first-found",
+		Felix: FelixConfig{
+			Metadataaddr:    "none",
+			Vxlanvni:        "4096",
+			MacPrefix:       "0E-2A",
+			LogSeverityFile: "none",
+			LogSeveritySys:  "none",
+		},
+		CNI: CalicoCNIConfig{
+			BinDir:   nodeConfig.AgentConfig.CNIBinDir,
+			ConfDir:  nodeConfig.AgentConfig.CNIConfDir,
+			IpamType: "calico-ipam",
+			Version:  "0.3.1",
+		},
 	}
 
-	cfg.CalicoConfig.KubeConfig = calicoKubeConfig
-	if err := getCNIConfigOverrides(cfg, hc); err != nil {
-		return nil, err
+	c.CalicoCNICfg.CalicoConfig.KubeConfig, err = c.createCalicoKubeConfig(ctx, restConfig)
+	if err != nil {
+		return err
 	}
 
-	agent.NetworkName = CalicoHnsNetworkName
+	return nil
+}
 
-	if err := createNodeNameFile(cfg, dataDir); err != nil {
-		return nil, err
+// writeConfigFiles writes the three required files by Calico
+func (c *Calico) writeConfigFiles(CNIConfDir string, NodeName string) error {
+	// parse the two config files required by Calico
+	templates, err := parseCalicoTemplatesFromConfig(c.CalicoCNICfg)
+	if err != nil {
+		return err
+	}
+	
+	// create calico kubeconfig, calico config and node name files
+	if err = util.WriteFile(filepath.Join("c:\\", c.DataDir, "agent", CalicoKubeConfigName), templates[calicoKubeConfigTemplate]); err != nil {
```

Comment:
We have a similar but different version of this at https://github.com/k3s-io/k3s/blob/master/pkg/daemons/control/deps/deps.go#L75 - I wonder if we could replace both of of these with something more streamlined that uses proper Kubernetes types and serialization instead of just rendering a string template?

Reply:
let me have a look and try to replicate the same